src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs

```diff
@@ -0,0 +1,424 @@
+                        // the user would need to clone to avoid the significant drop, just let
+                        // them know the issue exists so they can determine the best approach
+                        // for their use case.
+                        if found.lint_suggestion != MoveAndClone {
+                            diag.multipart_suggestion(
+                                found.suggestion_message(),
+                                vec![
+                                    (expr.span.shrink_to_lo(), replacement),
+                                    (found.found_span, found.scrutinee_replacement()),
+                                ],
+                                Applicability::MaybeIncorrect,
+                            );
+                            if let Some(diag_note) = found.diagnostic_note() {
+                                diag.note(diag_note);
+                            }
+                        }
+                    },
+                )
+            }
+        }
+    }
+}
+
+/// If the expression is an ExprKind::Match, check if the scrutinee has a significant drop that may
+/// have a surprising lifetime.
+fn has_significant_drop_in_scrutinee<'tcx, 'a>(
+    cx: &'a LateContext<'tcx>,
+    expr: &'tcx Expr<'tcx>,
+) -> Option<Vec<FoundSigDrop>> {
+    let mut helper = SigDropHelper::new(cx);
+    match expr.kind {
+        ExprKind::Match(match_expr, _, _) => helper.find_sig_drop(match_expr),
+        _ => None,
+    }
+}
+
+struct SigDropHelper<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    is_chain_end: bool,
+    seen_types: FxHashSet<Ty<'tcx>>,
+    has_significant_drop: bool,
+    current_sig_drop: Option<FoundSigDrop>,
+    sig_drop_spans: Option<Vec<FoundSigDrop>>,
+    special_handling_for_binary_op: bool,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum LintSuggestion {
+    MoveOnly,
+    MoveAndDerefToCopy,
+    MoveAndClone,
+}
+
+#[derive(Clone, Copy)]
+struct FoundSigDrop {
+    found_span: Span,
+    is_unit_return_val: bool,
+    lint_suggestion: LintSuggestion,
+}
+
+impl FoundSigDrop {
+    fn scrutinee_replacement(&self) -> String {
+        if self.is_unit_return_val {
+            "()".to_owned()
+        } else {
+            "value".to_owned()
+        }
+    }
+
+    fn create_replacement(&self, original: Cow<'_, str>, trailing_indent: String) -> String {
+        match self.lint_suggestion {
+            LintSuggestion::MoveOnly => {
+                if self.is_unit_return_val {
+                    format!("{};\n{}", original, trailing_indent)
+                } else {
+                    format!("let value = {};\n{}", original, trailing_indent)
+                }
+            }
+            LintSuggestion::MoveAndDerefToCopy => {
+                format!("let value = *{};\n{}", original, trailing_indent)
+            }
+            LintSuggestion::MoveAndClone => {
+                format!("let value = {}.clone();\n{}", original, trailing_indent)
+            }
```

Comment:
Excellent point! On further review, I decided to move the logic for creating a diagnostic suggestion into its own function so it could have an early return and consolidate some of the logic to avoid having to handle the `MoveAndClone` case in an awkward way.